### PR TITLE
[Merged by Bors] - feat: completion of the basic ShortComplex.Homology API

### PR DESCRIPTION
**Pull request by joelriou**
This PR marks the completion of the file `Algebra.Homology.ShortComplex.Homology`. It adds various API definitions and lemmas, including a comparison of the homology in a category and its opposite category.

---
- [x] depends on: #7192
- [x] depends on: #7193
- [x] depends on: #7194

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by riccardobrasca — COMMENTED** [review]
This file is getting quite long... do you plan to add a lot of other stuff? In this case we can maybe split it.

File: Mathlib/Algebra/Homology/ShortComplex/Homology.lean [hunk 1]
```diff
@@ -1013,6 +1013,160 @@ lemma π_homologyMap_ι [S₁.HasHomology] [S₂.HasHomology] (φ : S₁ ⟶ S
...
+/-- The natural isomorphism `(homologyFunctor C).op ≅ opFunctor C ⋙ homologyFunctor Cᵒᵖ`
+which relates the homology in `C` and in `Cᵒᵖ`. -/
+noncomputable def homologyFunctorOpNatIso [CategoryWithHomology C] :
+    (homologyFunctor C).op ≅ opFunctor C ⋙ homologyFunctor Cᵒᵖ :=
+  NatIso.ofComponents (fun S => S.unop.homologyOpIso.symm) (by simp)
+
+variable {C} {A : C}
+
+lemma liftCycles_homologyπ_eq_zero_of_boundary [S.HasHomology]
+    (k : A ⟶ S.X₂) (x : A ⟶ S.X₁) (hx : k = x ≫ S.f) :
```
[the remark this answers is not in the document]
Comment (joelriou) [reply Mathlib/Algebra/Homology/ShortComplex/Homology.lean:1057]: Making `x` implicit would make this lemma difficult to use. Here, we want to write `k` as the boundary of some `x`. Usually, the user would provide `x`, and in the best case scenario, `hx` is proven `by simp`.

[the remark this answers is not in the document]
Comment (riccardobrasca) [reply Mathlib/Algebra/Homology/ShortComplex/Homology.lean:1057]: Ah, I see.

File: Mathlib/Algebra/Homology/ShortComplex/Homology.lean [hunk 2]
```diff
@@ -1013,6 +1013,160 @@ lemma π_homologyMap_ι [S₁.HasHomology] [S₂.HasHomology] (φ : S₁ ⟶ S
...
+  rw [S.liftCycles_leftHomologyπ_eq_zero_of_boundary_assoc k x hx, zero_comp]
+
+@[reassoc]
+lemma homologyι_descOpcycles_π_eq_zero_of_boundary [S.HasHomology]
+    (k : S.X₂ ⟶ A) (x : S.X₃ ⟶ A) (hx : k = S.g ≫ x) :
+    S.homologyι ≫ S.descOpcycles k (by rw [hx, S.zero_assoc, zero_comp]) = 0 := by
+  dsimp only [homologyι]
+  rw [assoc, S.rightHomologyι_descOpcycles_π_eq_zero_of_boundary k x hx, comp_zero]
+
+lemma isIso_homologyMap_of_isIso_cyclesMap_of_epi (φ : S₁ ⟶ S₂)
```
[the remark this answers is not in the document]
Comment (joelriou) [reply Mathlib/Algebra/Homology/ShortComplex/Homology.lean]: Yes, thanks for the suggestion.

File: Mathlib/Algebra/Homology/ShortComplex/Homology.lean [hunk 3]
```diff
@@ -1013,6 +1013,160 @@ lemma π_homologyMap_ι [S₁.HasHomology] [S₂.HasHomology] (φ : S₁ ⟶ S
...
+      IsIso.hom_inv_id_assoc, toCycles_comp_homologyπ, comp_zero]
+  have ⟨z, hz⟩ := CokernelCofork.IsColimit.desc' S₂.homologyIsCokernel _ h
+  dsimp at hz
+  refine' ⟨⟨z, _, _⟩⟩
+  · rw [← cancel_epi S₁.homologyπ, homologyπ_naturality_assoc, hz,
+      IsIso.hom_inv_id_assoc, comp_id]
+  · rw [← cancel_epi S₂.homologyπ, reassoc_of% hz, homologyπ_naturality,
+      IsIso.inv_hom_id_assoc, comp_id]
+
+lemma isIso_homologyMap_of_isIso_opcyclesMap_of_mono (φ : S₁ ⟶ S₂)
```
[the remark this answers is not in the document]
Comment (joelriou) [reply Mathlib/Algebra/Homology/ShortComplex/Homology.lean]: (done)

File: Mathlib/Algebra/Homology/ShortComplex/Homology.lean [hunk 4]
```diff
@@ -1013,6 +1013,160 @@ lemma π_homologyMap_ι [S₁.HasHomology] [S₂.HasHomology] (φ : S₁ ⟶ S
...
+variable (S)
+
+/-- The canonical isomorphism `S.op.homology ≅ Opposite.op S.homology` when a short
+complex `S` has homology. -/
+noncomputable def homologyOpIso [S.HasHomology] :
+    S.op.homology ≅ Opposite.op S.homology :=
+  S.op.leftHomologyIso.symm ≪≫ S.leftHomologyOpIso ≪≫ S.rightHomologyIso.symm.op
+
+@[simp]
+lemma homologyMap'_op : (homologyMap' φ h₁ h₂).op =
```
[the remark this answers is not in the document]
Comment (joelriou) [reply Mathlib/Algebra/Homology/ShortComplex/Homology.lean:1024]: Thanks. I have just removed these `simp` attributes. (I had probably put these attributes in order to automatize a proof in the definition of `homologyFunctorOpNatIso`.)

**Comment by riccardobrasca** [comment]
LGTM, thanks!

bors d+

**Comment by joelriou** [comment]
Thanks @riccardobrasca for your suggestions, which I have mostly followed.

I do not intend to add any more material to this file. With this PR, the three large files `LeftHomology.lean`, `RightHomology.lean` and `Homology.lean` shall be exactly as they are in my dev branch https://github.com/leanprover-community/mathlib4/pull/4197, and these are the only three files that shall be >1000 lines long in this directory.

Next, I plan to develop a little bit more `PreservesHomology.lean` (following https://github.com/leanprover-community/mathlib4/pull/7197) before I return to `Exact.lean` (following https://github.com/leanprover-community/mathlib4/pull/7052).

**Comment by joelriou** [comment]
bors merge

**Pull request closed without merging**